pkg/argo_client/manifests.go

```diff
@@ -1,116 +1,439 @@
-
-	"github.com/zapier/kubechecks/telemetry"
+	"github.com/zapier/kubechecks/pkg"
+	"github.com/zapier/kubechecks/pkg/git"
+	"github.com/zapier/kubechecks/pkg/vcs"
 )
 
-func (argo *ArgoClient) GetManifestsLocal(ctx context.Context, name, tempRepoDir, changedAppFilePath string, app argoappv1.Application) ([]string, error) {
-	var err error
+type getRepo func(ctx context.Context, cloneURL string, branchName string) (*git.Repo, error)
 
-	ctx, span := tracer.Start(ctx, "GetManifestsLocal")
+func (a *ArgoClient) GetManifests(ctx context.Context, name string, app v1alpha1.Application, pullRequest vcs.PullRequest, getRepo getRepo) ([]string, error) {
+	ctx, span := tracer.Start(ctx, "GetManifests")
 	defer span.End()
 
-	log.Debug().Str("name", name).Msg("GetManifestsLocal")
+	log.Debug().Str("name", name).Msg("GetManifests")
 
 	start := time.Now()
 	defer func() {
 		duration := time.Since(start)
 		getManifestsDuration.WithLabelValues(name).Observe(duration.Seconds())
 	}()
 
-	clusterCloser, clusterClient := argo.GetClusterClient()
-	defer clusterCloser.Close()
+	contents, refs := a.preprocessSources(&app, pullRequest)
 
-	settingsCloser, settingsClient := argo.GetSettingsClient()
-	defer settingsCloser.Close()
+	var manifests []string
+	for _, source := range contents {
+		moreManifests, err := a.generateManifests(ctx, app, source, refs, pullRequest, getRepo)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to generate manifests")
+		}
+		manifests = append(manifests, moreManifests...)
+	}
+
+	getManifestsSuccess.WithLabelValues(name).Inc()
+	return manifests, nil
+}
+
+func (a *ArgoClient) preprocessSources(app *v1alpha1.Application, pullRequest vcs.PullRequest) ([]v1alpha1.ApplicationSource, []v1alpha1.ApplicationSource) {
+	if !app.Spec.HasMultipleSources() {
+		return []v1alpha1.ApplicationSource{app.Spec.GetSource()}, nil
+	}
+
+	// collect all ref sources, map by name
+	var contentSources []v1alpha1.ApplicationSource
+	var refSources []v1alpha1.ApplicationSource
+
+	for _, source := range app.Spec.Sources {
+		if source.Ref == "" {
+			contentSources = append(contentSources, source)
+			continue
+		}
+
+		if source.TargetRevision == pullRequest.BaseRef {
+			source.TargetRevision = pullRequest.HeadRef
+		}
+
+		refSources = append(refSources, source)
+	}
+
+	return contentSources, refSources
+}
+
+func (a *ArgoClient) generateManifests(ctx context.Context, app v1alpha1.Application, source v1alpha1.ApplicationSource, refs []v1alpha1.ApplicationSource, pullRequest vcs.PullRequest, getRepo func(ctx context.Context, cloneURL string, branchName string) (*git.Repo, error)) ([]string, error) {
+	// multisource apps must adhere to the following rules:
+	// 1. first source must be a non-ref source
+	// 2. there must be one and only one non-ref source
+	// 3. ref sources that match the pull requests's repo and target branch need to have their target branch swapped to the head branch of the pull request
```

Comment:
The point of kubechecks is to show you what will happen _after_ you merge the branch. We used to do that by cloning the repository and then merging the target branch into the base, and generating resources based on the result. As we're now relying on the repository service to do that work, we cannot manipulate the repositories. Instead, we manipulate the Application so that all references to the base git ref now point to the head ref. The end result should be the same.